File: app/src/main/java/com/coinninja/coinkeeper/cn/transaction/TransactionNotificationManager.java
package com.coinninja.coinkeeper.cn.transaction;

import com.coinninja.coinkeeper.model.TransactionNotificationMapper;
import com.coinninja.coinkeeper.model.db.Account;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.db.TransactionNotification;
import com.coinninja.coinkeeper.model.db.TransactionSummary;
import com.coinninja.coinkeeper.model.dto.CompletedBroadcastDTO;
import com.coinninja.coinkeeper.model.dto.CompletedInviteDTO;
import com.coinninja.coinkeeper.model.encryptedpayload.v1.TransactionNotificationV1;
import com.coinninja.coinkeeper.model.helpers.DaoSessionManager;
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient;
import com.coinninja.coinkeeper.service.client.model.Contact;
import com.coinninja.coinkeeper.util.CNLogger;
import com.coinninja.coinkeeper.util.PhoneNumberUtil;
import com.coinninja.coinkeeper.util.encryption.MessageEncryptor;

import javax.inject.Inject;

import retrofit2.Response;

public class TransactionNotificationManager {
    private final TransactionNotificationMapper transactionNotificationMapper;
    private final MessageEncryptor messageEncryptor;
    private final DaoSessionManager daoSessionManager;
    private final Account account;
    private final SignedCoinKeeperApiClient apiClient;
    private final CNLogger logger;

    @Inject
    TransactionNotificationManager(TransactionNotificationMapper transactionNotificationMapper,
                                   MessageEncryptor messageEncryptor, DaoSessionManager daoSessionManager,
                                   Account account, SignedCoinKeeperApiClient apiClient, CNLogger logger) {

        this.transactionNotificationMapper = transactionNotificationMapper;
        this.messageEncryptor = messageEncryptor;
        this.daoSessionManager = daoSessionManager;
        this.account = account;
        this.apiClient = apiClient;
        this.logger = logger;
    }

    public void saveTransactionNotificationLocally(InviteTransactionSummary inviteTransactionSummary, CompletedInviteDTO completedInviteDTO) {
        if (!completedInviteDTO.hasMemo()) return;

        TransactionNotification transactionNotification = createTransactionNotification(completedInviteDTO.getMemo(),
                completedInviteDTO.isMemoIsShared());

        inviteTransactionSummary.setTransactionNotification(transactionNotification);
        inviteTransactionSummary.update();
    }

    public void notifyCnOfFundedInvite(InviteTransactionSummary invite) {
        TransactionNotification transactionNotification = invite.getTransactionNotification();
        Contact contact = new Contact(invite.getReceiverPhoneNumber(), invite.getInviteName(), true);

        if (null == transactionNotification || null == invite.getPubkey()) {
            sendNotificationToCN(invite.getBtcTransactionId(), invite.getAddress(), contact.getHash(), "");
        } else {
            TransactionNotificationV1 message = transactionNotificationMapper.toV1(invite, account);
            String encryption = messageEncryptor.encrypt(message.toString(), invite.getPubkey());
            sendNotificationToCN(invite.getBtcTransactionId(), invite.getAddress(), contact.getHash(), encryption);
        }
    }

    public void saveTransactionNotificationLocally(TransactionSummary transactionSummary,
                                                   CompletedBroadcastDTO completedBroadcastActivityDTO) {
        if (!completedBroadcastActivityDTO.hasMemo()) return;

        TransactionNotification transactionNotification = createTransactionNotification(completedBroadcastActivityDTO.getMemo(),
                completedBroadcastActivityDTO.isMemoShared());
        transactionSummary.setTransactionNotification(transactionNotification);
        transactionNotification.setTxid(completedBroadcastActivityDTO.transactionId);
        transactionNotification.update();
        transactionSummary.update();
    }

    private TransactionNotification createTransactionNotification(String memo, boolean isShared) {
        TransactionNotification transactionNotification = daoSessionManager.newTransactionNotification();
        transactionNotification.setMemo(memo);
        transactionNotification.setIsShared(isShared);
        daoSessionManager.insert(transactionNotification);
        return transactionNotification;
    }

    public void sendTransactionNotificationToReceiver(CompletedBroadcastDTO completedBroadcastDTO) {
        TransactionNotificationV1 message = transactionNotificationMapper.toV1(completedBroadcastDTO, account);
        String encryption = messageEncryptor.encrypt(message.toString(), completedBroadcastDTO.getPublicKey());

        sendNotificationToCN(completedBroadcastDTO.transactionId,
                completedBroadcastDTO.getTransactionData().getPaymentAddress(),
                completedBroadcastDTO.getPhoneNumberHash(),
                encryption);
    }

    public void notifyOfPayment(CompletedBroadcastDTO completedBroadcastDTO) {
        sendNotificationToCN(completedBroadcastDTO.transactionId,
                completedBroadcastDTO.getTransactionData().getPaymentAddress(),
                completedBroadcastDTO.getPhoneNumberHash(),
                "");

    }

    private void sendNotificationToCN(String transactionId, String paymentAddress,
                                      String phoneNumberHash, String encryption) {

        Response response = apiClient.postTransactionNotification(transactionId,
                paymentAddress, phoneNumberHash, encryption);

        if (!response.isSuccessful()) {
            logger.logError(getClass().getSimpleName(), "Transaction Notification Post Failed", response);
        }
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/cn/transaction/TransactionNotificationManager.kt
package com.coinninja.coinkeeper.cn.transaction

import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.cn.transaction.notification.TransactionNotificationMapper
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary
import com.coinninja.coinkeeper.model.db.TransactionNotification
import com.coinninja.coinkeeper.model.db.TransactionSummary
import com.coinninja.coinkeeper.model.db.UserIdentity
import com.coinninja.coinkeeper.model.dto.CompletedBroadcastDTO
import com.coinninja.coinkeeper.model.dto.CompletedInviteDTO
import com.coinninja.coinkeeper.model.helpers.DaoSessionManager
import com.coinninja.coinkeeper.model.helpers.DropbitAccountHelper
import com.coinninja.coinkeeper.model.helpers.UserIdentityHelper
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import com.coinninja.coinkeeper.service.client.model.CNSharedMemo
import com.coinninja.coinkeeper.util.CNLogger
import com.coinninja.coinkeeper.util.encryption.MessageEncryptor
import javax.inject.Inject

@Mockable
class TransactionNotificationManager @Inject
internal constructor(internal val transactionNotificationMapper: TransactionNotificationMapper,
                     internal val messageEncryptor: MessageEncryptor,
                     internal val daoSessionManager: DaoSessionManager,
                     internal val apiClient: SignedCoinKeeperApiClient,
                     internal val dropbitAccountHelper: DropbitAccountHelper,
                     internal val userIdentityHelper: UserIdentityHelper,
                     internal val logger: CNLogger) {

    fun saveTransactionNotificationLocally(inviteTransactionSummary: InviteTransactionSummary, completedInviteDTO: CompletedInviteDTO) {
        if (!completedInviteDTO.hasMemo()) return

        val transactionNotification = createTransactionNotification(completedInviteDTO.memo,
                completedInviteDTO.isMemoIsShared)

        inviteTransactionSummary.transactionNotification = transactionNotification
        inviteTransactionSummary.update()
    }

    fun notifyCnOfFundedInvite(invite: InviteTransactionSummary) {
        val transactionNotification = invite.transactionNotification

        if (null == transactionNotification || null == invite.pubkey) {
            sendNotificationToCN(invite.btcTransactionId, invite.address, invite.toUser.hash, "")
        } else {
            val message = transactionNotificationMapper.toEncryptionMessage(invite)
            val encryption = messageEncryptor.encrypt(message, invite.pubkey)
            sendNotificationToCN(invite.btcTransactionId, invite.address, invite.toUser.hash, encryption)
        }
    }

    fun saveTransactionNotificationLocally(transactionSummary: TransactionSummary,
                                           completedBroadcastActivityDTO: CompletedBroadcastDTO) {
        if (!completedBroadcastActivityDTO.hasMemo()) return

        val transactionNotification = createTransactionNotification(completedBroadcastActivityDTO.memo,
                completedBroadcastActivityDTO.isMemoShared)
        transactionSummary.transactionNotification = transactionNotification
        transactionNotification.txid = completedBroadcastActivityDTO.transactionId
        completedBroadcastActivityDTO.identity?.let { toIdentity ->
            val toUser: UserIdentity = userIdentityHelper.updateFrom(toIdentity)
            transactionNotification.toUser = toUser
            val fromAccount = dropbitAccountHelper.profileForIdentity(toUser)
            fromAccount?.let {
                transactionNotification.fromUser =  userIdentityHelper.updateFrom(fromAccount)
            }
        }
        transactionNotification.update()
        transactionSummary.update()
    }

    fun createTransactionNotification(memo: String?, isShared: Boolean): TransactionNotification {
        val transactionNotification = daoSessionManager.newTransactionNotification()
        transactionNotification.memo = memo
        transactionNotification.isShared = isShared
        daoSessionManager.insert(transactionNotification)
        return transactionNotification
    }

    fun sendTransactionNotificationToReceiver(completedBroadcastDTO: CompletedBroadcastDTO) {
        val message = transactionNotificationMapper.toEncryptionMessage(completedBroadcastDTO)
        val encryption = messageEncryptor.encrypt(message, completedBroadcastDTO.publicKey)

        sendNotificationToCN(completedBroadcastDTO.transactionId,
                completedBroadcastDTO.transactionData.paymentAddress ?: "",
                completedBroadcastDTO.identity?.hashForType ?: "",
                encryption)
    }

    private fun sendNotificationToCN(transactionId: String?, paymentAddress: String,
                                     identityHash: String, encryption: String) {

        val sharedMemo = CNSharedMemo(transactionId, paymentAddress, identityHash, encryption)
        val response = apiClient.postTransactionNotification(sharedMemo)

        if (!response.isSuccessful) {
            logger.logError(javaClass.simpleName, "Transaction Notification Post Failed", response)
        }
    }

}
